### ISSUE TITLE
FR: Progress bar - reverse order for RTL books

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#7917]
**Does your feature request involve difficulty completing a task? Please describe.**
I am not sure how difficult this could be to implement.
This is not a major inconvenience or anything, so I'd understand if you say no.
I just think this would make reading RTL books more consistent in terms of user experience.

So manga is usually read from right to left, top to bottom. Actual physical books are opened from the "end".
In KOReader there is an option `["inverse_reading_order"] = true` which flips the tap zones and swipes for page turns alright.
So it would be nice if the progress bar (at least the one on the bottom, since it is common for both engines) was drawn in the reverse order as well.

**Describe alternatives you've considered**
Otherwise, manga works great as epub files in the usual LTR order.

**Additional context**
I think this might be actually useful for other RTL documents types as well (Arabic, Hebrew?). I just don't happen to actually read RTL documents other than manga.

Thank you for your consideration!

**Comment by poire-z** [comment]
The footer is considered part of the KOReader frontend UI, so it follows the UI language LTR/RTL property.
With an Arabic or Hebrew UI, you would get the footer laid out from right to left.

You can see how it would look laid out from right to left (the footer, but also the menu and everything else) with:
`File browser > Tools > More tools > Developer options > UI layout mirroring and text direction > [x] Reverse UI layout mirroring`
and I'm sure you'll forget the idea soon ;)
(To me, it looks more like a cultural thing - a RTL footer might only be usable to arabic/hebrew readers who has this hardcoded in their brain :)

**Comment by mergen3107** [comment]
@poire-z 
Can I enable this in Tweak documents settings (by directories)? 



> only be usable to arabic/hebrew readers who has this hardcoded in their brain
Actually, the way manga is read forces me to "think" RTL as well :D So I wouldn't mid to have RTL UI while reading it.

**Comment by poire-z** [comment]
No, it's global and tweaks many things - so it needs a restart to have all these things cleaned out when toggling it.

**Comment by mergen3107** [comment]
OK, I see. Thank you! :)
I guess I should drop my hopes for a possibility to make only progress bar reversed?

**Comment by poire-z** [comment]
Dunno. Depends on how other RTL manga readers would like the footer reversed - could be just a new footer setting to follow reading page order, and someone willing to implement it.
But I would at least expect you to live with the reverse UI while reading a few mangas (if it's that immersive, you might bear ok the ugly reversed menus) and see how resilient the RTL part of your mind really is :)

**Comment by poire-z** [comment]
^But then the footer might not be enough: you'd want the Skim widget to also show the progress bar revese, and probably the ToC to be RTL too so you can browse it RTL - and so bookmarks too...
If you're fine with the menus being reversed, a small user-plugin that if enabled would detect if there's a mismatch between the inverse_reading_order doc settings and the UI layout direction global setting, and fix the latter, and restart KOReader, might work.

**Comment by mergen3107** [comment]
> you'd want the Skim widget to also show the progress bar revese
ah, yes, I completely forgot about it. However I haven't used it yet.

As for ToC and Bookmarks I think they should stay the same, bacause RTL or not, they should follow top-to-bottom order anyway, right? I mean Chapter 2 comes after Chapter 1 regardless of RTL.

> If you're fine with the menus being reversed
Yes, it is bearable - I will need them just to find manga files and open them (to fall into a new world then).

**Comment by hius07** [comment]
If change to `true`, is it what you need?
https://github.com/koreader/koreader/blob/c0386e5dd02014794e4715528b6e8ae82f287aba/frontend/ui/widget/progresswidget.lua#L52

**Comment by mergen3107** [comment]
Yes, exactly! :) This draws both progress bar and Skim widget RTL. Thank you so much! 
I can at least set this manually now. 
It still requires a restart to apply so I could probably create a shell script to replace the file (or better search and replace one line to make compatible with updates) to run before reading manga

**Comment by mergen3107** [comment]
[RTL_progressbar.zip](https://github.com/koreader/koreader/files/6749305/RTL_progressbar.zip)
So I created two simple scripts, to find `fill_from_right` and replace `false,` to `true,` in that line, and reverse.
I put the files into my manga directory, run ON script, open the manga book, restart KOReader. 
Thank you guys!

**Comment by mergen3107** [comment]
For completeness, I also swap the direction of swipes in `koreader/settings/gestures.lua` like this:

```
sed -i -e 's/one_finger_swipe_bottom_edge_left/one_finger_swipe_bottom_edge_leeft/' gestures.lua.bak 2>_error.txt
sed -i -e 's/one_finger_swipe_bottom_edge_right/one_finger_swipe_bottom_edge_left/' gestures.lua.bak 2>_error.txt
sed -i -e 's/one_finger_swipe_bottom_edge_leeft/one_finger_swipe_bottom_edge_right/' gestures.lua.bak 2>_error.txt
```

Works good both ways (no pun intended) so far :)

**Comment by mergen3107** [comment]
The swipes above are for next and previous chapter. So “swipe along the reading direction” is next chapter, “swipe against the reading direction” is previous chapter, for RTL as well (with the script)

**Comment by Commodore64user** [comment]
okay, I don't really have any horses in this race but, I have taken a slight interest in this. I have gone for a things-with-progress-bar inversion only approach, that is to say inverting the bar in the footer and the skim widget, everything else stays the same as the original UI layout.

**Comment by mergen3107** [comment]
> exactly one year

I am not a robot I promise 🤖

**Comment by mergen3107** [comment]
> I don't really have any horses in this race but, I have taken a slight interest in this

(Weird, first part of my reply above is gone)
Hooray, that statistically usually means  I will have something to test next week :D thank you!

**Comment by mergen3107** [comment]
Re-opening to the let PR close it

**Comment by hius07** [comment]
> sorting of actions in Profiles

Edit actions - 2nd page of the menu

**Comment by mergen3107** [comment]
Thanks! I found it

**Comment by mergen3107** [comment]
Well, great intermediate success!

1) I enabled Book shortcuts plugin. 
2) Top menu - Tools - Book shortcuts - New shortcut.
Selected this script named `_RTL_chapter_swipes_TOGGLE.sh`:
```
#!/bin/sh
cd /mnt/us/koreader/settings/
sed -i -e 's/one_finger_swipe_bottom_edge_left/one_finger_swipe_bottom_edge_leeft/' gestures.lua 2>_error.txt
sed -i -e 's/one_finger_swipe_bottom_edge_right/one_finger_swipe_bottom_edge_left/' gestures.lua 2>_error.txt
sed -i -e 's/one_finger_swipe_bottom_edge_leeft/one_finger_swipe_bottom_edge_right/' gestures.lua 2>_error.txt
```
My `gestures.lua` has these bottom swipes from chapter jumping (along the progress bar) in the reader:
```
        ["one_finger_swipe_bottom_edge_right"] = {
            ["prev_chapter"] = true,
            ["settings"] = {
                ["order"] = {
                    [1] = "prev_chapter",
                },
            },
        },
        ["one_finger_swipe_bottom_edge_left"] = {
            ["next_chapter"] = true,
            ["settings"] = {
                ["order"] = {
                    [1] = "next_chapter",
                },
            },
        },
```
(and something else in File Manager, but that doesn't matter)

3) Created a test Profile `[manga] toggle Chapter swipes`. 
Auto Exec: on book opening and book closing with `[manga]` in the path (all my manga books are in `[manga]` folder).
Action: General - `_RTL_chapter_swipes_TOGGLE.sh` (this is the Book Shortcut, even though it doesn't say so - be careful how you name your scripts :D ) Thanks @hius07 for #13393 
I also added `Restart KOReader` but it seems this is not necessary. See below.
4) Now when I open a book from `[manga]`, it swaps the left/right bottom swipes for chapter jumps.
With @Commodore64user's #13382, the tweak above and my directory defaults to swap the taps/swipes as well (https://koreader.rocks/user_guide/#manga) I now have :
- Inverted taps/swipes for RTL as in manga
- Inverted RTL progress bar
- Inverted Chapter swipe navigation!

When I exit the book from `[manga]`, it reverts back!

It seems like `Auto exec` in particular kicks in early in the book so that KOReader restart is not requires to have changes in the `gestures.lua` to be applied after book opening.
OR `Restart KOReader` action didn't work at all?
OR `gestures.lua` changes don't need restart at all anymore (compare to the beginning of this ticket)?

Anyway, my next steps are:
1) Test when opening another `[manga]` books from within successfully RTL-swapped `[manga]` book. I suppose since the switch is relative, chapter jump swap will only work every other book :D 
2) Come up with a better idea on how to definitively swap the `gestures.lua` file. Maybe, make two permanent copies: `gestures_RTL.lua` and `gestures_LTR.lua` and have those to overwrite `gestures.lua`, with two separate profiles? One for book opening, another for book closing with `[manga]` in the path. However, this will wear out the storage a bit.

Any better ideas are welcome! :D

**Comment by mergen3107** [comment]
> OR `Restart KOReader` action didn't work at all?

@hius07 
Two findings with `Auto exec` on book opening with path:
1) With this order:
- Run script
- Restart KOReader

`Run script` ends with `Script exited successfully` - a notification which stays on screen until tap.

2) If I only leave `Restart KOReader` - I get stuck in a loop :D 
- open book
- KOReader restarts - another auto exec kicks in, asking to restart
- KOReader restarts - another auto exec kicks in, asking to restart...

I had to `killall reader.lua` from ssh to stop it and then edited `profiles.lua`

**Comment by mergen3107** [comment]
> 2\. Come up with a better idea on how to definitively swap the `gestures.lua` file. Maybe, make two permanent copies

Interestingly enough, this didn't work. So, file rename doesn't apply, and on-the-fly file change with `sed` does? By "works" I mean KOReader accepts those new changes with `sed` without a restart...

**Comment by mergen3107** [comment]
How does `Taps and Gestures` apply changes without the restart?
Can the same mechanism added to the Dispatcher?
I know it now sounds complicated:
- Have a profile
- that has an action in Dispatcher
- that changes a gesture for one specific gesture with one specific new action...

**Comment by hius07** [comment]
I think it's a bad practice to change internal lua files on the fly.
Many (most/all) of them are rewritten by the program on exiting.

**Comment by mergen3107** [comment]
What's the better solution then?

**Comment by hius07** [comment]
To reassign a gesture - action? I think we do not have the method outside of Taps and Gestures menu so far.

**Comment by mergen3107** [comment]
I think for now I can resort to outside-KOReader methods. 

Winterbreak allows running custom scripts. I can make two scripts that rename files and then start KOReader. 

Not great, but the best I can now :D 

Maybe, a user patch that simulates setting the taps and gestures?

**Comment by Commodore64user** [comment]
instead of doing gymnastics, why don't you set both `one_finger_swipe_bottom_edge_right` and left to `next chapter` and `one_finger_swipe_top_edge_right` and left to `previous chapter` or whatever. _Not ideal_ (I know) but at least that solves the problem of needing to invert them.

**Comment by mergen3107** [comment]
My original idea is this:
- Swipe along the direction of increasing pages (0 to 100%) on the progress bar: next chapter
- Swipe against this direction: prev chapter
- I like KOReader's progress bar more than Alt Status bar because it is common for mupdf/cre docs, and has more options
- This is why I stick with bottom edge for swipes
- This is way RTL and LTR directions are opposite 
- I use top edge for swipe both in reader and FM (left swipe: previous doc, right swipe: History). I wouldn't like to change my habits :D 

If I can't make it work, that's fine. I am just trying as hard as I can to see if I can get it to work. 

Can progress bar hijack bottom gestures?
I am thinking of new gestures to setup:
- a checkbox that says "Override bottom swipe gestures with progress bar gestures" (also the same for bottom corner taps, etc?)
- New gestures "Swipe along progress bar direction" and "Swipe against progress bar direction", where along/against are determined by "Invert Progress Bar" setting

**Comment by mergen3107** [comment]
Just for fun, checked the method I proposed before :D 
1) Created copies of `gestures.lua` file twice: for RTL and for LTR.
2) Created two scripts to launch KOReader (possible after WinterBreak)
3) Profit!

It is inconvenient, so for now I just don't swap the gestures at all. But it is safe, and it works if I really need to use it :D 

This is the closest I ever got solving this issue.

[WinterBreak_KOReader_Scripts_Swap_Chapter_Swipes.zip](https://github.com/user-attachments/files/19217432/WinterBreak_KOReader_Scripts_Swap_Chapter_Swipes.zip)

**Comment by hosamn** [comment]
This is frustrating, Can someone please explain to me why is this (Reverse Progress bar direction for RTL books) not implemented by default when user activates the option (document direction RTL) or when a RTL book is detected?!

**Comment by mergen3107** [comment]
Waiting for #13382 to be merged. 

But it won't be by default. You will need to add a profile that sets this setting in RTL books.

**Comment by hosamn** [comment]
Ok, but WHY is it a separate option?! That does not make sense, does it?!, why is it not applied with the option (document direction RTL)?!

Here is a scenario, the user opens an RTL book, the sytem recognizes that it's RTL (either automatically from metadata or from the user manually activating the document direction RTL option) then the progress bar gets reversed.

Does this scenario make sense?!

**Comment by mergen3107** [comment]
Please stop your "?!" nonsense. I appreciate your contribution in discussion, but please refrain from emotional elevation. 

Let's discuss. 

1) Upcoming PR is only a settings. Users can decide to make it run on certain books. Definitive way to do so is to target books with certain path (e.g., I selected books with path containing "manga").

2) "RTL detection" is not a thing, as far as I know.

3) At least in my use case, I also have swipe gestures which directions coincident with the direction of the progress bar progression. So I wouldn't always like to have RTL progress bar on "RTL books" because I want to preserve my gestures alignment.

**Comment by poire-z** [comment]
We nowhere detect if a document is RTL.
`Writing direction RTL` in the bottom menu (if that's what you are talking about) is a PDF/CBZ only option, and only instruct the PDF renderer how to do its "reflow" feature.

**Comment by rukit24** [comment]
I have downloaded and copy the latest Commodore64user change file commit, but in Koreader UI. Nothing showing. Did i miss something, or what do i need to do to tricker the RTL progressbar
